library.c

```diff
@@ -2043,6 +2043,24 @@ redis_sock_set_err(RedisSock *redis_sock, const char *msg, int msg_len)
     }
 }
 
+PHP_REDIS_API int
+redis_sock_set_stream_context(RedisSock *redis_sock, zval *options)
+{
+    zend_string *zkey;
+    zval *z_ele;
+
+    if (!redis_sock || Z_TYPE_P(options) != IS_ARRAY) {
+        return FAILURE;
+    } else if (!redis_sock->stream_ctx) {
+        redis_sock->stream_ctx = php_stream_context_alloc();
+    }
+    ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(options), zkey, z_ele) {
```

Comment:
I wonder if we should add logic to throw a warning if the user specifies `local_cert`, `local_pk` or `cafile` to files that don't exist.  Right now it will silently fail but the user won't know exactly why.

Something like this maybe?  I'm not sure if there are other options that point to files though.
```patch
diff --git library.c library.c
index fcbb4fd..75232a3 100644
--- library.c
+++ library.c
@@ -35,6 +35,7 @@
 #endif
 
 #include <ext/standard/php_rand.h>
+#include <Zend/zend_virtual_cwd.h>
 
 #define UNSERIALIZE_NONE 0
 #define UNSERIALIZE_KEYS 1
@@ -2046,7 +2047,8 @@ redis_sock_set_err(RedisSock *redis_sock, const char *msg, int msg_len)
 PHP_REDIS_API int
 redis_sock_set_stream_context(RedisSock *redis_sock, zval *options)
 {
-    zend_string *zkey;
+    char fullpath[MAXPATHLEN];
+    zend_string *zkey, *ztest;
     zval *z_ele;
 
     if (!redis_sock || Z_TYPE_P(options) != IS_ARRAY) {
@@ -2054,7 +2056,20 @@ redis_sock_set_stream_context(RedisSock *redis_sock, zval *options)
     } else if (!redis_sock->stream_ctx) {
         redis_sock->stream_ctx = php_stream_context_alloc();
     }
+
     ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(options), zkey, z_ele) {
+        if (ZSTR_STRICMP_STATIC(zkey, "local_cert") ||
+            ZSTR_STRICMP_STATIC(zkey, "local_pk") ||
+            ZSTR_STRICMP_STATIC(zkey, "cafile"))
+        {
+            ztest = zval_get_string(z_ele);
+            if (!VCWD_REALPATH(ZSTR_VAL(ztest), fullpath)) {
+                php_error_docref(NULL, E_WARNING, "Can't find file '%s' to configure %s", ZSTR_VAL(ztest), ZSTR_VAL(zkey));
+                continue;
+            }
+            zend_string_release(ztest);
+        }
+
         php_stream_context_set_option(redis_sock->stream_ctx, "ssl", ZSTR_VAL(zkey), z_ele);
     } ZEND_HASH_FOREACH_END();
```

Reply:
@michael-grunder I used tests script from https://github.com/phpredis/phpredis/issues/1600#issuecomment-634901555 and I see the error about invalid cafile

```
php -d extension=modules/redis.so issue-1600.php 
PHP Warning:  failed loading cafile stream: `/root/redis-6.0.3/tests/tls/ca.crt' in /home/pyatsukhnenko/code/phpredis/issue-1600.php on line 5
PHP Warning:  Redis::connect(): Failed to enable crypto in /home/pyatsukhnenko/code/phpredis/issue-1600.php on line 5
PHP Fatal error:  Uncaught RedisException: Redis server went away in /home/pyatsukhnenko/code/phpredis/issue-1600.php:11
Stack trace:
#0 /home/pyatsukhnenko/code/phpredis/issue-1600.php(11): Redis->set('test_tls', 'it works')
#1 {main}
  thrown in /home/pyatsukhnenko/code/phpredis/issue-1600.php on line 11
```

Reply:
![300px-Surprised_Pikachu](https://user-images.githubusercontent.com/468149/83575853-42003f80-a4e5-11ea-808b-3152aadb22a0.jpg)

OK let's just merge it as-is and I'll open a PR with validation :joy: